runtime(vimscript): add simple vimscript complete function

habamax opened the pull request:
I have come up with a simple omnicomplete function for vimscript to complete different things (commands, functions, options, global/buffer variables, expressions, highlight command).

It works nicely with the `set autocomplete`. While it is not super smart, it might be better than nothing and as usual there is a room for future improvements.

Set `setlocal complete^=o^10` if you would like to use it with `set autocomplete` or just trigger it manually with `C-x C-o`:

[![asciicast](https://asciinema.org/a/730271.svg)](https://asciinema.org/a/730271)

@chrisbra, @girishji, @lifepillar, @yegappan could you please check if it works for you? And in general if it makes sense to put it into vim's core.

habamax committed:
add simple vimscript complete function

habamax committed:
better `set option option=value,value2` complete

girishji commented:
You may be missing some entries from `'cmdline'`. Take a look at this: https://github.com/girishji/vimcomplete/blob/main/tools/vimdictgen.vim

habamax commented:
> You may be missing some entries from `'cmdline'`. Take a look at this: https://github.com/girishji/vimcomplete/blob/main/tools/vimdictgen.vim

Thx, will check it.

habamax committed:
set omnifunc in vim's ftplugin

Shougo commented:
I think `getcompletiontype()` is better to parse Vim script.

habamax commented:
> I think `getcompletiontype()` is better to parse Vim script.

I tried it but it can't do `&option` as in `echo getcompletiontype("&option")` is a `command` which is not what I want to complete. So I went full regex parsing instead.

Maybe it could be used in some parts of the code, let me see.

habamax committed:
use getcompletiontype() as a fallback

habamax committed:
fix comment spelling

habamax committed:
use either fuzzy or stridx depending on 'completeopt'

Fix CTRL-X CTRL-O by filtering augmented items too.

habamax committed:
simplify completion logic: mostly rely on getcompletion

use getcompletion(..., 'cmdline') to get proper completion details

habamax committed:
complete commands and functions

getcompletiontype() returns 'command' for both commands and vim9script functions

e.g. winrestview() --> is a 'command'

Use 2 getcompletion() calls for functions and commands.

brianhuster commented:
> The difference with CTRL-X CTRL-V is that omnicomplete function works with set autocomplete provided you have added setlocal complete+=o to your filetype. And of course the fact that we can improve completion way easier using vimscript instead of getting into C source code.

So do you plan to make `Ctrl-x Ctrl-v` just call this omnifunc?

habamax committed:
re-add complete in interpolated vim string

Shougo commented:
> I tried it but it can't do &option as in echo getcompletiontype("&option") is a command which is not what I want to complete. So I went full regex parsing instead.

Yes.  But you can use the pattern.  `getcompletiontype("echo &op")` works.
You can use the full command line to parse.

habamax commented on runtime/ftplugin/vim.vim:
```diff
@@ -116,6 +116,9 @@ setlocal include=\\v^\\s*import\\s*(autoload)?
 setlocal define=\\v^\\s*export\\s*(def\|const\|var\|final)
 
+" set omnifunc completion
+setlocal omnifunc=vimcomplete#Complete
```
Can’t they just not get this in the first place?

habamax replied:
I disagree with that, I believe we should target vim in vim project.

But, well...

habamax replied:
And we don’t have the vim9script guard here https://github.com/vim/vim/blob/6a6a44ffb5ad7350a4d313a5faeae17e50bf37a2/runtime/ftplugin/c.vim#L32

chrisbra replied:
Then let's fix it there as well :) (I'll do it when merging this). Thanks all

chrisbra replied:
So how about this then:
```diff
diff --git a/CONTRIBUTING.md b/CONTRIBUTING.md
index 8f0947acf..9e187f3d1 100644
--- a/CONTRIBUTING.md
+++ b/CONTRIBUTING.md
@@ -48,7 +48,7 @@ approvers are happy with that particular change.

 # Reporting issues

-We use GitHub issues, but that is not a requirement. Writing to the Vim
+We use GitHub [issues][16], but that is not a requirement. Writing to the Vim
 mailing list is also fine.

 Please use the GitHub issues only for actual issues. If you are not 100% sure
@@ -72,7 +72,8 @@ Or open [the todo file][todo list] on GitHub to see the latest version.

 The latest version of these files can be obtained from the repository.
 They are usually not updated with numbered patches. However, they may
-or may not work with older Vim releases (since they may contain new features).
+or may not work with older Vim releases (since they may depend on new
+features).

 If you find a problem with one of these files or have a suggestion for
 improvement, please first try to contact the maintainer directly.
@@ -82,7 +83,18 @@ upstream repository. You may also check the [MAINTAINERS][11] file.
 The maintainer will take care of issues and send updates to the Vim project for
 distribution with Vim.

-If the maintainer does not respond, contact the [vim-dev][0] mailing list.
+If the maintainer does not respond, contact the [vim-dev][0] mailing list or
+open an [issue][16] here.
+
+Note: Whether or not to use Vim9 Script is up to the maintainer. For runtime
+files maintained here, we aim to preserve compatibility with Neovim if
+possible. Please add wrap Vim9 script with a guard like this:
+```vim
+if has('vim9script')
+   " use Vim9 script implementation
+   [...]
+endif
+```

 ## Contributing new runtime files

@@ -95,6 +107,9 @@ PR with your changes against this repository here. For new filetypes, do not for
 - add yourself as Maintainer to the top of file (again, keep the header similar to
   other runtime files)
 - add yourself to the [MAINTAINERS][11] file.
+- add a guard `if has('vim9script')` if you like to maintain Neovim
+  compatibility but want to use Vim9 script (or restrict yourself to legacy Vim
+  script)

 # Translations

@@ -144,3 +159,4 @@ mailing list. For other questions please use the [Vi Stack Exchange][8] website,
 [14]: https://github.com/vim/vim/blob/master/runtime/doc/syntax.txt
 [15]: https://en.wikipedia.org/wiki/Developer_Certificate_of_Origin
 [16]: https://github.com/vim/vim/blob/master/runtime/doc/helphelp.txt
+[16]: https://github.com/vim/vim/issues
```

bfrg replied:
`vim9script` can be disabled when compiling Vim from source. Therefore, the check `has('vim9script')` is necessary, even in Vim itself. Though, not sure if `runtime` would work when `vim9script` is disabled since a few functions in `runtime` have been ported to `vim9script`.

zeertzjq replied:
Look at version.c line 630. It always shows `+vim9script` (even if `FEAT_EVAL` is disabled).

brianhuster replied:
@chrisbra  Given that Vim9script is always available in Vim9 built with `+eval`, I don't think Vim should add `has("vim9script")` guard anywhere like that. It also doesn't help Nvim either, because from what I see in [Nvim development guide](https://neovim.io/doc/user/dev_vimpatch.html#dev-vimpatch-not-applicable), any patches related to Vim9script are just marked as NA, which mean they are just not ported to Nvim (and looks like [they haven't ported this patch either](https://github.com/neovim/neovim/blob/master/runtime/ftplugin/vim.vim))

Also if the point is to sync ftplugin with Nvim, then adding `has("vim9script")` to `set omnifunc` in C ftplugin would not make sense either, because the `ccomplete` script in Nvim is still written in legacy Vimscript.

chrisbra replied:
Well, I don't know if such a change to the vim ftplugin would be marked as "Vim9script", however this is done and then 
it also does not hurt, plus it helps when a user with Vim8 simply downloads the update plugin files from here (not sure if this is a valid use case). 

In any case, that is the best I can do to maintain compatibility with Neovim.

habamax commented:
thanks!

Shane-XB-Qian commented:
> For me this is a surprise that we are to think about neovim compatibility.

yes, unknown this how come. :smile:
// from last time: "let g:vimsyn_vim_features = ['nvim']" ?
// it is weird, not sure Debian Whether setup something for Ubuntu, or firefox setup something for chrome :lol:

-- 
shane.xb.qian
